<div id="container" style="position:relative;">
<div style="float:left"><h1> Capstone Project: Diabetes risk predictor based on health indicators</h1></div>
<div style="float:left"><h2> Notebook 1: Data Loading and Cleaning - Created by Diego Villanueva</h2></div>

</div>
</div>

This notebook contains the first section fo the BrainStation Capstone project, a prediction tool to identify individuals with a higher risk of developing Diabetes or Pre-diabetes based on basic health indicators 

The section includes loading and cleaning of the data that will be used for models in the following notebooks.


### Data Dictionary

<table>
    <tr>
        <th>Column</th>
        <th>Data Type</th>
        <th>Description</th>
        <th>Possible values</th>
    </tr>
    <tr>
        <td>TARGET VARIABLE: Diabetes_012</td>
        <td>float64</td>
        <td>Does the participant have diabetes or prediabetes</td>
        <td>0 = no diabetes 1 = prediabetes 2 = diabetes</td>
    </tr>
    <tr>
        <td>HighBP</td>
        <td>float64</td>
        <td>Has the participant being told by a medical professional that they have high blood pressure</td>
        <td>0 = no high BP 1 = high BP</td>
    </tr>
    <tr>
        <td>HighChol</td>
        <td>float64</td>
        <td>Has the participant being told by a medical professional that they have high blood cholesterol</td>
        <td>0 = no high cholesterol 1 = high cholesterol</td>
    </tr>
    <tr>
        <td>CholCheck</td>
        <td>float64</td>
        <td>Has the participant cholesterol been checked in the last 5 years</td>
        <td>0 = no cholesterol check in 5 years 1 = yes cholesterol check in 5 years</td>
    </tr>
    <tr>
        <td>BMI</td>
        <td>float64</td>
        <td>The Body Mass Index (BMI) of the participant</td>
        <td>Any number from 12 to 98</td>
    </tr>
    <tr>
        <td>Smoker</td>
        <td>float64</td>
        <td>Has the participant smoked at least 100 cigarretes in their entire life</td>
        <td>0 = no 1 = yes</td>
    </tr>
    <tr>
        <td>Stroke</td>
        <td>float64</td>
        <td>Has the participant ever had a stroke</td>
        <td>0 = no 1 = yes</td>
    </tr>
    <tr>
        <td>HeartDiseaseorAttack</td>
        <td>float64</td>
        <td>Has the participant has had a Coronary Heart Disease (CHD) or myocardial infarction (MI)</td>
        <td>0 = no 1 = yes</td>
    </tr>
    <tr>
        <td>PhysActivity</td>
        <td>float64</td>
        <td>Has the participant done any physical activity in past 30 days</td>
        <td>0 = no 1 = yes</td>
    </tr>
    <tr>
        <td>Fruits</td>
        <td>float64</td>
        <td>Does the participant consume at least 1 fruit per day</td>
        <td>0 = no 1 = yes</td>
    </tr>
    <tr>
        <td>Veggies</td>
        <td>float64</td>
        <td>Does the participant consume vegetables at least once per day</td>
        <td>0 = no 1 = yes</td>
    </tr>
    <tr>
        <td>HvyAlcoholConsump</td>
        <td>float64</td>
        <td>Does the participant consume 14 drinks or more per week if they are an adult male, or 7 drinks or more per week if they are an adult female</td>
        <td>0 = no 1 = yes</td>
    </tr>
    <tr>
        <td>AnyHealthcare</td>
        <td>float64</td>
        <td>Does the participant have any kind of health care coverage (e.g., health insurance, prepaid plans)</td>
        <td>0 = no 1 = yes</td>
    </tr>
    <tr>
        <td>NoDocbcCost</td>
        <td>float64</td>
        <td>Has the participant had the need to to see a doctor in the past 12 months but could not do it because of financial reasons?</td>
        <td>0 = no 1 = yes</td>
    </tr>
    <tr>
        <td>GenHlth</td>
        <td>float64</td>
        <td>How does the participant describe their health levels on a scale from 1 to 5 </td>
        <td>1 = excellent 2 = very good 3 = good 4 = fair 5 = poor</td>
    </tr>
    <tr>
        <td>MentHlth</td>
        <td>float64</td>
        <td>How many days of bad mental health has the participant had in the last 30 days</td>
        <td>1 to 30</td>
    </tr>
    <tr>
        <td>PhysHlth</td>
        <td>float64</td>
        <td>How many days has the participant experienced physical illness or an injury in the last 30 days</td>
        <td>1 to 30</td>
    </tr>
    <tr>
        <td>DiffWalk</td>
        <td>float64</td>
        <td>Does the participant have serious difficulty walking or climbing stairs</td>
        <td>0 = no 1 = yes</td>
    </tr>
    <tr>
        <td>Sex</td>
        <td>float64</td>
        <td>What gender was the participant assigned at birth </td>
        <td>0 = female 1 = male</td>
    </tr>
    <tr>
        <td>Age</td>
        <td>float64</td>
        <td>Group age of the participant</td>
        <td>1 to 13: 1 = 18 to 24, 2 = 25 to 29, 3 = 30 to 34, 4 = 35 to 39, 5 = 40 to 44, 6 = 45 to 49, 7 = 50 to 54, 8 = 55 to 59, 9 = 60 to 64, 10 = 65 to 69, 11 = 70 to 74, 12 = 75 to 79, 13 = 80+</td>
    </tr>
    <tr>
        <td>Education</td>
        <td>float64</td>
        <td>The maximum level of education of the participant. Grouped in 6 different education levels.</td>
        <td>1 = Never attended school or only kindergarten, 2 = Elementary education, 3 = Some high school, 4 = High school graduate, 5 = College 1 year to 3 years,  6 = College graduate and above</td>
    </tr>
    <tr>
        <td>Income</td>
        <td>float64</td>
        <td>Desc</td>
        <td>1 to 8: 1 = less than 10,000, 2 = 10,001 to 20,000, 3 = 20,001 to 30,000, 4 = 30,001 to 40,000, 5 = 40,001 to 50,000, 6 = 50,001 to 60,000, 7 = 60,001 to 70,000, 8 = 70,001 or more </td>
    </tr>
</table>

### Imports

In [4]:
# data manipulation
import numpy as np
import pandas as pd

# plotting
import matplotlib.pyplot as plt
import seaborn as sns

### Read csv file and preliminary data exploration

In [5]:
# Load csv file
df_raw = pd.read_csv("../data/Raw data/Diabetes_Health_Indicators_Raw.csv")

In [14]:
# Check data
df_raw.head()

Diabetes_012  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0           0.0     1.0       1.0        1.0  40.0     1.0     0.0   
1           0.0     0.0       0.0        0.0  25.0     1.0     0.0   
2           0.0     1.0       1.0        1.0  28.0     0.0     0.0   
3           0.0     1.0       0.0        1.0  27.0     0.0     0.0   
4           0.0     1.0       1.0        1.0  24.0     0.0     0.0   

   HeartDiseaseorAttack  PhysActivity  Fruits  ...  AnyHealthcare  \
0                   0.0           0.0     0.0  ...            1.0   
1                   0.0           1.0     0.0  ...            0.0   
2                   0.0           0.0     1.0  ...            1.0   
3                   0.0           1.0     1.0  ...            1.0   
4                   0.0           1.0     1.0  ...            1.0   

   NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex   Age  Education  \
0          0.0      5.0      18.0      15.0       1.0  0.0   9.0        4.0   
1          1.0      3.0       0.0       0.0       0.0  0.0   7.0        6.0   
2          1.0      5.0      30.0      30.0       1.0  0.0   9.0        4.0   
3          0.0      2.0       0.0       0.0       0.0  0.0  11.0        3.0   
4          0.0      2.0       3.0       0.0       0.0  0.0  11.0        5.0   

   Income  
0     3.0  
1     1.0  
2     8.0  
3     6.0  
4     4.0  

[5 rows x 22 columns]

In [19]:
data_columns = df_raw.shape[0]
data_rows = df_raw.shape[1]
print(f'The dataframe contains {data_columns} columns and {data_rows} rows')

The dataframe contains 253680 columns and 22 rows


### Datatypes and formats

In [15]:
# Generate overview of dataset variables
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 253680 entries, 0 to 253679
Data columns (total 22 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Diabetes_012          253680 non-null  float64
 1   HighBP                253680 non-null  float64
 2   HighChol              253680 non-null  float64
 3   CholCheck             253680 non-null  float64
 4   BMI                   253680 non-null  float64
 5   Smoker                253680 non-null  float64
 6   Stroke                253680 non-null  float64
 7   HeartDiseaseorAttack  253680 non-null  float64
 8   PhysActivity          253680 non-null  float64
 9   Fruits                253680 non-null  float64
 10  Veggies               253680 non-null  float64
 11  HvyAlcoholConsump     253680 non-null  float64
 12  AnyHealthcare         253680 non-null  float64
 13  NoDocbcCost           253680 non-null  float64
 14  GenHlth               253680 non-null  float64
 15  

In [9]:
df_raw.describe().T

count       mean       std   min   25%   50%   75%  \
Diabetes_012          253680.0   0.296921  0.698160   0.0   0.0   0.0   0.0   
HighBP                253680.0   0.429001  0.494934   0.0   0.0   0.0   1.0   
HighChol              253680.0   0.424121  0.494210   0.0   0.0   0.0   1.0   
CholCheck             253680.0   0.962670  0.189571   0.0   1.0   1.0   1.0   
BMI                   253680.0  28.382364  6.608694  12.0  24.0  27.0  31.0   
Smoker                253680.0   0.443169  0.496761   0.0   0.0   0.0   1.0   
Stroke                253680.0   0.040571  0.197294   0.0   0.0   0.0   0.0   
HeartDiseaseorAttack  253680.0   0.094186  0.292087   0.0   0.0   0.0   0.0   
PhysActivity          253680.0   0.756544  0.429169   0.0   1.0   1.0   1.0   
Fruits                253680.0   0.634256  0.481639   0.0   0.0   1.0   1.0   
Veggies               253680.0   0.811420  0.391175   0.0   1.0   1.0   1.0   
HvyAlcoholConsump     253680.0   0.056197  0.230302   0.0   0.0   0.0   0.0   
AnyHealthcare         253680.0   0.951053  0.215759   0.0   1.0   1.0   1.0   
NoDocbcCost           253680.0   0.084177  0.277654   0.0   0.0   0.0   0.0   
GenHlth               253680.0   2.511392  1.068477   1.0   2.0   2.0   3.0   
MentHlth              253680.0   3.184772  7.412847   0.0   0.0   0.0   2.0   
PhysHlth              253680.0   4.242081  8.717951   0.0   0.0   0.0   3.0   
DiffWalk              253680.0   0.168224  0.374066   0.0   0.0   0.0   0.0   
Sex                   253680.0   0.440342  0.496429   0.0   0.0   0.0   1.0   
Age                   253680.0   8.032119  3.054220   1.0   6.0   8.0  10.0   
Education             253680.0   5.050434  0.985774   1.0   4.0   5.0   6.0   
Income                253680.0   6.053875  2.071148   1.0   5.0   7.0   8.0   

                       max  
Diabetes_012           2.0  
HighBP                 1.0  
HighChol               1.0  
CholCheck              1.0  
BMI                   98.0  
Smoker                 1.0  
Stroke                 1.0  
HeartDiseaseorAttack   1.0  
PhysActivity           1.0  
Fruits                 1.0  
Veggies                1.0  
HvyAlcoholConsump      1.0  
AnyHealthcare          1.0  
NoDocbcCost            1.0  
GenHlth                5.0  
MentHlth              30.0  
PhysHlth              30.0  
DiffWalk               1.0  
Sex                    1.0  
Age                   13.0  
Education              6.0  
Income                 8.0

### Preliminary Data Exploration Observations
Upon initial examination, the dataframe has 253,680 columns and 22 rows.

All columns consist of numeric values with a float64 data type. We may need to convert some columns to an integer data type during the pre-processing phase. The necessity for this will be assessed after a detailed analysis of each feature.

There's no unique identifier assigned to each survey participant.

Furthermore, every column is free from null values.

### Duplicate values

In [32]:
duplicate_rows = df_raw.duplicated().sum()
duplicate_perc = round(duplicate_rows/df_raw.shape[0]*100, 1)
print(f'The dataset contains a total of {duplicate_rows} duplicate rows. This represents {duplicate_perc}% of the total rows in the dataframe')


The dataset contains a total of 23899 duplicate rows. This represents 9.4% of the total rows in the dataframe


The dataframe contains 23,899 repeated rows. Without a unique identifier for each survey participant, it's challenging to determine whether rows with identical data are genuine duplicates or simply represent participants with similar health conditions.

This data is sourced from the Behavioral Risk Factor Surveillance System (BRFSS) conducted by the Centers for Disease Control and Prevention. And it is safe to beleive that the CDC adheres to a rigorous data collection process designed to minimize errors.

Given the possibility that individuals may share similar health conditions, I will keep these rows until there's stronger evidence to suggest they are indeed duplicates.

### Missing data

In [37]:
df_cleaned.isna().sum()

Diabetes_012            0
HighBP                  0
HighChol                0
CholCheck               0
BMI                     0
Smoker                  0
Stroke                  0
HeartDiseaseorAttack    0
PhysActivity            0
Fruits                  0
Veggies                 0
HvyAlcoholConsump       0
AnyHealthcare           0
NoDocbcCost             0
GenHlth                 0
MentHlth                0
PhysHlth                0
DiffWalk                0
Sex                     0
Age                     0
Education               0
Income                  0
dtype: int64

The results above show that there are no missing values in the dataframe.

Therefore, I will save a new file with the cleaned dataframe and proceed with the EDA in the following notebook.

### Fixing data types and column names
Before I start making changes to the dataframe, I will create a copy of it.


In [18]:
df_cleaned = df_raw.copy()
df_cleaned.head().T

0     1     2     3     4
Diabetes_012           0.0   0.0   0.0   0.0   0.0
HighBP                 1.0   0.0   1.0   1.0   1.0
HighChol               1.0   0.0   1.0   0.0   1.0
CholCheck              1.0   0.0   1.0   1.0   1.0
BMI                   40.0  25.0  28.0  27.0  24.0
Smoker                 1.0   1.0   0.0   0.0   0.0
Stroke                 0.0   0.0   0.0   0.0   0.0
HeartDiseaseorAttack   0.0   0.0   0.0   0.0   0.0
PhysActivity           0.0   1.0   0.0   1.0   1.0
Fruits                 0.0   0.0   1.0   1.0   1.0
Veggies                1.0   0.0   0.0   1.0   1.0
HvyAlcoholConsump      0.0   0.0   0.0   0.0   0.0
AnyHealthcare          1.0   0.0   1.0   1.0   1.0
NoDocbcCost            0.0   1.0   1.0   0.0   0.0
GenHlth                5.0   3.0   5.0   2.0   2.0
MentHlth              18.0   0.0  30.0   0.0   3.0
PhysHlth              15.0   0.0  30.0   0.0   0.0
DiffWalk               1.0   0.0   1.0   0.0   0.0
Sex                    0.0   0.0   0.0   0.0   0.0
Age                    9.0   7.0   9.0  11.0  11.0
Education              4.0   6.0   4.0   3.0   5.0
Income                 3.0   1.0   8.0   6.0   4.0

In [22]:
# Change the names of columns to names that provide better guidance on the data in the column
#Change the name of Diabetes_012 column
df_cleaned.rename(columns={'Diabetes_012': 'Has_diabetes'}, inplace=True)

#Change the name of  HighBP column
df_cleaned.rename(columns={'HighBP': 'Has_high_bp'}, inplace=True)

#Change the name of  HighChol column
df_cleaned.rename(columns={'HighChol': 'Has_high_cholesterol'}, inplace=True)

#Change the name of  CholCheck column
df_cleaned.rename(columns={'CholCheck': 'Checked_cholesterol'}, inplace=True)

#Change the name of  Smoker column
df_cleaned.rename(columns={'Smoker': 'Is_smoker'}, inplace=True)

#Change the name of  Stroke column
df_cleaned.rename(columns={'Stroke': 'Had_stroke'}, inplace=True)

#Change the name of  HeartDiseaseorAttack column
df_cleaned.rename(columns={'HeartDiseaseorAttack': 'Has_heart_disease'}, inplace=True)

#Change the name of  PhysActivity column
df_cleaned.rename(columns={'PhysActivity': 'Physically_active'}, inplace=True)

#Change the name of  Fruits column
df_cleaned.rename(columns={'Fruits': 'Eats_fruits'}, inplace=True)

#Change the name of  Veggies column
df_cleaned.rename(columns={'Veggies': 'Eats_veggies'}, inplace=True)

#Change the name of  HvyAlcoholConsump column
df_cleaned.rename(columns={'HvyAlcoholConsump': 'Heavy_drinker'}, inplace=True)

#Change the name of  AnyHealthcare column
df_cleaned.rename(columns={'AnyHealthcare': 'Has_healthcare_cov'}, inplace=True)

#Change the name of  NoDocbcCost column
df_cleaned.rename(columns={'NoDocbcCost': 'No_attention_bc_cost'}, inplace=True)

#Change the name of  GenHlth column
df_cleaned.rename(columns={'GenHlth': 'General_health_score'}, inplace=True)

#Change the name of  MentHlth column
df_cleaned.rename(columns={'MentHlth': 'Mental_health_bad_days'}, inplace=True)

#Change the name of  PhysHlth column
df_cleaned.rename(columns={'PhysHlth': 'Physical_health_bad_days'}, inplace=True)

#Change the name of  DiffWalk column
df_cleaned.rename(columns={'DiffWalk': 'Walking_difficulty'}, inplace=True)

#Change the name of  Sex column
df_cleaned.rename(columns={'Sex': 'Is_male'}, inplace=True)

df_cleaned.head().T

0     1     2     3     4
Has_diabetes               0.0   0.0   0.0   0.0   0.0
Has_high_bp                1.0   0.0   1.0   1.0   1.0
Has_high_cholesterol       1.0   0.0   1.0   0.0   1.0
Checked_cholesterol        1.0   0.0   1.0   1.0   1.0
BMI                       40.0  25.0  28.0  27.0  24.0
Is_smoker                  1.0   1.0   0.0   0.0   0.0
Had_stroke                 0.0   0.0   0.0   0.0   0.0
Has_heart_disease          0.0   0.0   0.0   0.0   0.0
Physical_activity          0.0   1.0   0.0   1.0   1.0
Eats_fruits                0.0   0.0   1.0   1.0   1.0
Eats_veggies               1.0   0.0   0.0   1.0   1.0
Heavy_drinker              0.0   0.0   0.0   0.0   0.0
Has_healthcare_cov         1.0   0.0   1.0   1.0   1.0
No_attention_bc_cost       0.0   1.0   1.0   0.0   0.0
General_health_score       5.0   3.0   5.0   2.0   2.0
Mental_health_bad_days    18.0   0.0  30.0   0.0   3.0
Physical_health_bad_days  15.0   0.0  30.0   0.0   0.0
Walking_difficulty         1.0   0.0   1.0   0.0   0.0
Is_male                    0.0   0.0   0.0   0.0   0.0
Age                        9.0   7.0   9.0  11.0  11.0
Education                  4.0   6.0   4.0   3.0   5.0
Income                     3.0   1.0   8.0   6.0   4.0

In [23]:
df_cleaned.to_csv("../Data/Cleaned data/Diabetes_Health_Indicators_Cleaned_.csv", index=False)